### ISSUE TITLE
Require unique block names in Binaryen IR?

### ISSUE THREAD

**Issue by kripken** [WebAssembly/binaryen#735]
The s-expression language allows e.g.

````
(block $x
  (block $x
    ..
  )
) 
````

This is convenient for writing code, but when optimizing the IR it has the downside that seeing a block and a nested break with the same name is not a guarantee that the former breaks to the latter. If names were unique we would have that guarantee. That would get rid of some overhead in the optimizer, make things simpler, and avoid some possible bugs (I just ran into such a one).

The downside is that the s-expression parser would need to unique-ify names when they do collide, which means that the loaded module doesn't look exactly the same as the s-expression file (arguably, though, in some ways it makes it more readable). In other words, just like binaryen IR not supporting stacky stuff (#663), this would make a round-trip conversion differ slightly in a corner case. Seems worth it to me. Thoughts?

**Comment by dschuff** [comment]
TBH I'm not super thrilled about string names being the only/primary way to refer to so many things in Binaryen IR. Obviously that was a design decision and not a bug, and it does have some advantages, but it leads to difficulty and bugs like these. (Also it means that the IR can faithfully represent valid wasm without any names, although obviously that doesn't have to be a first-class goal now) If I were doing it from scratch I'd probably have pointers or some other handle be the way objects refer to each other in the IR, with name lookup as an optional addition. Obviously that's a bigger change and maybe you don't want that anyway :) But given that names don't need to be unique in the input language you either have to uniquify them or deal with the possibility of collisions (I suppose to be sure of break targets in the current IR you'll have to have code in several places that walks the scopes outward rather than relying on the names). Given that 1:1 is already not a goal and the IR design relies pretty heavily on names being the primary handle for so many things, making them unique is probably the way to go (unless you do want to get a little less reliant on the names).

**Comment by kripken** [comment]
Ok, cool.

Yeah, using names is possibly a controversial design decision. But overall:

 * For global objects, names allow the module to be created and optimized in parallel (add a call node to a function not yet existing, for example).
 * Names are interned, so they are fast as pointers to compare. They are sort of like handles, really.
 * With that said, for branch names - the issue here - I agree things are more debatable. I think numbers are too fragile, e.g. `br 2` could break if you add a block in the middle. But pointers which you suggested might be interesting to consider (depending on overhead, debuggability, etc.).

**Comment by kripken** [comment]
This has been implemented in that merged PR.
